06_gpu_and_ml/audio-to-text/parakeet.py

```diff
@@ -0,0 +1,238 @@
+        "nvidia/cuda:12.8.0-cudnn-devel-ubuntu22.04", add_python="3.12"
+    )
+    .pip_install("uv")
+    .env(
+        {
+            "HF_HUB_ENABLE_HF_TRANSFER": "1",
+            "HF_HOME": "/cache",  # cache directory for Hugging Face models
+            "DEBIAN_FRONTEND": "noninteractive",
+            "CXX": "g++",
+            "CC": "g++",
+        }
+    )
+    .apt_install("ffmpeg")
+    .run_commands(
+        "uv pip install --system hf_transfer huggingface_hub[hf-xet] nemo_toolkit[asr] cuda-python>=12.3",
+        "uv pip install --system 'numpy<2.0'",
+        "uv pip install --system fastapi",
+    )
+)
+
+with image.imports():
+    import numpy as np
+    import nemo.collections.asr as nemo_asr
+
+# ## Implementing real-time audio transcription on Modal
+
+# Now we're ready to implement the transcription model. We wrap inference in a Modal [Cls](https://modal.com/docs/guide/lifecycle-functions) that
+# ensures models are loaded and then moved to the GPU once when a new container starts.
+
+# The `load` method loads the model at start, instead of during inference, using [`modal.enter()`](https://modal.com/docs/reference/modal.enter#modalenter).
+# The `transcribe` method takes a numpy array of audio data, and returns the transcribed text.
+# The `web` method creates a FastAPI app using [`modal.asgi_app`](https://modal.com/docs/reference/modal.asgi_app#modalasgi_app) that serves a
+# [websocket](https://modal.com/docs/guide/webhooks#websockets) endpoint for real-time audio transcription.
+
+# Additionally, since this app is running in its own container, we make a `.remote` call to invoke
+# the `transcribe` method. This allows us to use the GPU for inference.
+
+
+@app.cls(volumes={"/cache": model_cache}, gpu="a10g", image=image)
+class Parakeet:
+    @modal.enter()
+    def load(self):
+        self.model = nemo_asr.models.ASRModel.from_pretrained(
+            model_name="nvidia/parakeet-tdt-0.6b-v2"
+        )
+
+    @modal.method()
+    def transcribe(self, audio_data: np.ndarray) -> str:
+        output = self.model.transcribe([audio_data])
+        return output[0].text
+
+    @modal.asgi_app()
+    def web(self):
+        from fastapi import FastAPI, WebSocket, WebSocketDisconnect, Response
+
+        web_app = FastAPI()
+
+        @web_app.get("/status")
+        async def status():
+            return Response(status_code=200)
+
+        @web_app.websocket("/ws")
+        async def websocket_endpoint(ws: WebSocket):
+            await ws.accept()
+            buffer = bytearray()
+
+            try:
+                while True:
+                    chunk = await ws.receive_bytes()
+                    buffer.extend(chunk)
+
+                    if len(buffer) > BUFFER_SIZE:
+                        audio_np = np.frombuffer(bytes(buffer), dtype=np.int16)
+                        buffer.clear()
+
+                        try:
+                            text = self.transcribe.remote(audio_np)
```

Comment:
Calling `text = self.transcribe.remote(audio_np)` from this websocket will call this in a _different_  container. This can be fine if expect a container to work as a dispatcher and one to do the GPU processing. That doesn't seem to be the case here, though. Use `text = self.transcribe.local(audio_np)` instead.

Reply:
got it, ty!